Add postfix convention for store naming rule

ilyaryabchinski opened the pull request:
Adds a possibility to configure convention for store naming. Prefix convention will be a default.
Most new files are created for testing purposes 🙂 

closes #31

igorkamyshev commented:
Thanks for your contribution 🙏

I'm little bit busy now, so I'll review it on Monday.

ilyaryabchinski commented:
Don't worry 👌 Have a great rest of the week.

igorkamyshev commented on rules/enforce-store-naming-convention/enforce-store-naming-convention.js:
```diff
@@ -38,11 +45,24 @@ module.exports = {
           const storeName = node.id.name;
 
-          if (storeName?.startsWith("$")) {
+          if (mode === "prefix" && storeName?.startsWith("$")) {
```
Shall we extract this logic to some place? I see, it is used here and in the JS-case. Also, I suppose it should be used in `effector/no-getState` JS-case (it detects stores by naming convention).

ilyaryabchinski replied:
I think in this case we have to change the implementation and use [shared settings](https://eslint.org/docs/user-guide/configuring/configuration-files#adding-shared-settings). Otherwise, it wouldn’t be possible to share the mode value across several rules.

What do you think?

igorkamyshev replied:
It seems like the good idea 👍

igorkamyshev commented on rules/enforce-store-naming-convention/enforce-store-naming-convention.js:
```diff
@@ -38,11 +45,24 @@ module.exports = {
+            return;
+          }
+
+          const correctedStoreName = mode === "prefix"
```
I suggest a move defining of `correctedStoreName` to report-function because it is related only to suggestions.

igorkamyshev commented on rules/enforce-store-naming-convention/enforce-store-naming-convention.js:
```diff
@@ -74,22 +94,35 @@ module.exports = {
         if (node.callee?.property?.name === "map") {
-          const objectIsEffectorStore =
-            node.callee?.object?.name?.startsWith?.("$");
+          const objectIsEffectorStore = mode === "prefix"
```
As a mention before, this logic can be extracted to a separate function, so this piece of code could be simplified.

igorkamyshev commented on rules/enforce-store-naming-convention/enforce-store-naming-convention.js:
```diff
@@ -101,14 +134,24 @@ module.exports = {
 
           const storeName = node.parent.id.name;
-          if (storeName.startsWith("$")) {
+          if (mode === "prefix" && storeName.startsWith("$")) {
```
Same same.

igorkamyshev commented on rules/enforce-store-naming-convention/enforce-store-naming-convention.js:
```diff
@@ -125,14 +168,24 @@ module.exports = {
 
           const storeName = node.parent.id.name;
-          if (storeName.startsWith("$")) {
+          if (mode === "prefix" && storeName.startsWith("$")) {
```
And here

igorkamyshev commented on rules/enforce-store-naming-convention/enforce-store-naming-convention.js:
```diff
@@ -141,19 +194,24 @@ module.exports = {
     messageId: "invalidName",
     data: {
       storeName,
+      correctedStoreName,
```
I think, this is the better place to define `correctedStoreName`.

igorkamyshev commented on rules/enforce-store-naming-convention/enforce-store-naming-convention.md:
````diff
@@ -1,11 +1,39 @@
+```js
+module.exports = {
+  rules: {
+    "effector/enforce-store-naming-convention": ["error", "prefix"],
````
I thought, `prefix` is default option. Could we omit it here?

igorkamyshev commented on rules/enforce-store-naming-convention/postfix/examples/correct-store-naming.js:
```diff
@@ -0,0 +1,17 @@
+"use strict";
```
Looks like compiled code. Are you confident that is a good example?

igorkamyshev commented on rules/enforce-store-naming-convention/enforce-store-naming-convention.js:
```diff
@@ -7,19 +7,26 @@ module.exports = {
-    const parserServices = context.parserServices;
+    const { parserServices, options } = context;
+    // prefix mode is a default option
+    const [mode = "prefix"] = options;
```
May we validate input here? What if user write incorrect mode? I think, we should raise a readable exception in this case.

ilyaryabchinski replied:
I added this, but not sure if the message is clear 😄 Do you have something on your mind?

igorkamyshev replied:
I like the message, but I think, we should add more information — like "the configuration of effector-plugin-eslint ..."

igorkamyshev reviewed (changes requested):
Thanks for your contribution, one more time. It's a great help.

I found some issues in this PR. Could you fix it, please?

Also, I found a logic problem of this feature. The rule `effector/no-getState` in JS-case (with unavailable type information) rely on naming convention to detect store. I think, we should add a support of postfix-notation in this rule too. Can we get options of this rule from one another? Or, we can use both notations to detect stores (it can lead to false-positive triggers in some cases)?

ilyaryabchinski commented:
Thank you for the review!👍

I feel sick today but I will try to fix all the comments on the upcoming weekend or the latest at the beginning of the next week.

I hope it doesn't affect your plans for 1.0 version of the project 🙏

igorkamyshev commented:
Sure, it's ok.

Ne boley 🍀

ilyaryabchinski commented:
Seems like I fixed all those things you mentioned above, but now I have a couple of questions 😄 :

1.  Do you think that we also need to throw an error in the rule `effector/no-getState` if the convention is configured in the wrong way? Or it's enough to throw only in the rule `effector/enforce-store-naming-convention`?
2. I noticed that now store names like `$myStore$ are valid. Do you think we should cover these cases as well` ?

igorkamyshev commented on rules/enforce-store-naming-convention/enforce-store-naming-convention.js:
```diff
@@ -141,21 +160,40 @@ module.exports = {
   });
 }
+
+function validateStoreNameConvention(storeNameConvention) {
```
So, this function looks like shred validation utility, which must be used in any rule with store defining.

igorkamyshev commented on rules/enforce-store-naming-convention/enforce-store-naming-convention.md:
````diff
@@ -1,11 +1,44 @@
 // 👎 bad name
 const name = createStrore(null);
 ```
+## Postfix convention
````
Shall we move documentation about shred settings to main README? It seems like general info.

ilyaryabchinski replied:
I think yes it's a good idea to copy this info to README, but I'd also still keep the information here since it's related to the particular rule

igorkamyshev commented on utils/is-store-name-valid.js:
```diff
@@ -0,0 +1,13 @@
+function isStoreNameValid(storeName, storeNameConvention) {
```
Maybe replace `storeNameConvention` by `context`? In this case, we can move validation here too, and simplify code of rules.

igorkamyshev reviewed (commented):
Sorry for the delay 🥲

It's a great job, thanks! I left some minor comments. I think, it's the last iteration 🌚

igorkamyshev commented:
@ilyaryabchinski also, in my opinion, `$myStore$` is a bad way to name stores.

igorkamyshev commented:
@ilyaryabchinski I'm going to release 0.3.0 at the weekend. If you'll finish this PR in a few days, I can wait for you before it. What do you think?

ilyaryabchinski commented:
I am gonna work on your comments this weekend, but not sure if I can finalize all the stuff by the end of the weekend. so I think it would be safer if you release 0.3.0 without these changes.

Sorry for the delays I'm having quite busy times at work due to the end of Q3 ⏳

igorkamyshev commented:
It's okay, no worries.

ilyaryabchinski commented:
Your vision is great, Thank you for your comments! I resolved them and added some logic to handle [these kinds of edge cases](https://github.com/effector/eslint-plugin/pull/37#issuecomment-923628542). I also extracted several functions to `utils` so now it looks a little cleaner(well, at least to me 😄 ). 

Let's see if we are finally done with this issue 😸 

![5o97qv](https://user-images.githubusercontent.com/20639767/134786532-e197f199-2930-4c79-b62c-0b895f47f6ce.jpeg)

igorkamyshev reviewed (approved):
Good job! Thanks a lot, it'a great help 🙏 

Sorry for such a huge number of comments 😢 

It'll be released it 0.3.0 today later.

ilyaryabchinski commented:
Don't worry, I understand u have some vision and ideas on the project and also it was my first experience developing custom ESLint rules, so I learned a lot. Thank you for the collaboration 😊👍
